[TSP Migration] Migrate healthbot from swagger to TypeSpec

welovej opened the pull request:
This PR migrates your latest version (identified by the tag in your readme.md) of swagger to TypeSpec. We already tried our best to make sure the TypeSpec represents same as previous swagger. Since we lack the business knowledge, please validate this PR again to make sure it's functional equivalent as before. The local validation step is at [Getting started | TypeSpec Azure](https://azure.github.io/typespec-azure/docs/migrate-swagger/01-get-started/#review-and-adjust-the-typespec)

Besides, TypeSpec encourages to follow [ARM guidelines](https://aka.ms/rpguidelines). Therefore, some representations in your previous swagger will be fixed to follow these guidelines. When you see differences in your local validation, please keep this note in mind. Here is the list of your anti-pattern:
- Your resource definition is not referring the resource definitions from common-type.

Please reach out to [TypeSpec Discussions Channel](https://teams.microsoft.com/l/channel/19%3a906c1efbbec54dc8949ac736633e6bdf%40thread.skype/TypeSpec%2520Discussion%2520%25F0%259F%2590%25AE?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) if there is any help needed.

welovej committed:
after being converted

welovej committed:
tsp optimized and compiled into swagger

welovej committed:
removed too early tsp rename and type fix

pshao25 commented on specification/healthbot/resource-manager/readme.md:
````diff
@@ -37,7 +37,7 @@ These settings apply only when `--tag=package-2024-02` is specified on the comma
 ```yaml $(tag) == 'package-2024-02'
 input-file:
-  - Microsoft.HealthBot/stable/2024-02-01/healthbot.json
+  - Microsoft.HealthBot/stable/2024-02-01/openapi.json
````
Keep its original name.

Petermarcu commented on specification/healthbot/resource-manager/Microsoft.HealthBot/stable/2024-02-01/openapi.json:
Is this replacing the healthbot.json? Should git mv have been used to track the file rename so the diff will show as well?

welovej replied:
Reverted

welovej committed:
Update readme.md

welovej committed:
fix page

welovej committed:
Update tspconfig.yaml

welovej committed:
Update tspconfig.yaml

welovej committed:
update

welovej committed:
update

welovej committed:
update

Commit:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into TspMig-healthbot

welovej committed:
Update tspconfig.yaml

welovej committed:
Merge branch 'TspMig-healthbot' of https://github.com/welovej/azure-rest-api-specs into TspMig-healthbot

MaryGao commented on specification/healthbot/Healthbot.Management/tspconfig.yaml:
```diff
@@ -0,0 +1,47 @@
+    namespace: "com.microsoft.azure.management.healthbot"
+    service-name: "HealthBot" # human-readable service name, whitespace allowed
+    flavor: azure
+  "@azure-tools/typespec-ts":
```
```suggestion
  "@azure-tools/typespec-ts":
      service-dir: sdk/healthbot
```

MaryGao committed:
Update specification/healthbot/Healthbot.Management/tspconfig.yaml

MaryGao commented on specification/healthbot/Healthbot.Management/tspconfig.yaml:
```diff
@@ -0,0 +1,48 @@
+    service-name: "HealthBot" # human-readable service name, whitespace allowed
+    flavor: azure
+  "@azure-tools/typespec-ts":
+      service-dir: sdk/healthbot
```
```suggestion
    service-dir: sdk/healthbot
```

MaryGao committed:
Update specification/healthbot/Healthbot.Management/tspconfig.yaml

welovej committed:
prettier

welovej committed:
Update tspconfig.yaml

welovej committed:
prettier exmaples

welovej committed:
Delete diff.json

welovej committed:
Update models.tsp

welovej committed:
Update tspconfig.yaml

welovej committed:
Update back-compatible.tsp

welovej committed:
Update main.tsp

Commit:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into TspMig-healthbot

Commit:
update

welovej committed:
Merge branch 'TspMig-healthbot' of https://github.com/welovej/azure-rest-api-specs into TspMig-healthbot

welovej committed:
Update healthbot.json

JiaqiZhang-Dev commented:
/azp run

JiaqiZhang-Dev committed:
Merge branch 'main' into TspMig-healthbot

JiaqiZhang-Dev commented on specification/healthbot/Healthbot.Management/HealthBot.tsp:
```diff
@@ -0,0 +1,160 @@
+}
+
+@armResourceOperations
+interface HealthBots {
```
Should we rename the interface name to `Bots`？

v-jiaodi committed:
Merge branch 'main' into TspMig-healthbot

v-jiaodi committed:
resolve breaking change

v-jiaodi committed:
Merge branch 'main' into TspMig-healthbot

v-jiaodi committed:
resolve breaking change with client.tsp

welovej committed:
FIXME update

welovej committed:
Merge remote-tracking branch 'upstream/main' into TspMig-healthbot

welovej committed:
example files

welovej committed:
update

welovej committed:
recoery of previous api version

msyyc commented on specification/healthbot/Healthbot.Management/client.tsp:
```diff
@@ -0,0 +1,5 @@
+import "./main.tsp";
+using Azure.ClientGenerator.Core;
+
+@@clientName(Microsoft.HealthBot, "HealthbotClient", "javascript");
```
```suggestion
@@clientName(Microsoft.HealthBot, "HealthbotClient", "javascript");
@@clientName(Microsoft.HealthBot, "HealthBotMgmtClient", "python");
```

welovej committed:
fix for model validation

v-jiaodi committed:
Merge branch 'main' into TspMig-healthbot

v-jiaodi committed:
Merge branch 'main' into TspMig-healthbot

weidongxu-microsoft commented on specification/healthbot/Healthbot.Management/tspconfig.yaml:
```diff
@@ -0,0 +1,51 @@
+    flavor: "azure"
+  "@azure-tools/typespec-java":
+    package-dir: "azure-resourcemanager-healthbot"
+    namespace: "com.microsoft.azure.management.healthbot"
```
```suggestion
    namespace: "com.azure.resourcemanager.healthbot"
```

weidongxu-microsoft committed:
Update specification/healthbot/Healthbot.Management/tspconfig.yaml

weidongxu-microsoft commented on specification/healthbot/Healthbot.Management/tspconfig.yaml:
```diff
@@ -0,0 +1,51 @@
+  "@azure-tools/typespec-java":
+    package-dir: "azure-resourcemanager-healthbot"
+    namespace: "com.azure.resourcemanager.healthbot"
+    service-name: "HealthBot" # human-readable service name, whitespace allowed
```
```suggestion
    service-name: "Healthbot" # human-readable service name, whitespace allowed
```

weidongxu-microsoft committed:
Update specification/healthbot/Healthbot.Management/tspconfig.yaml

Commit:
update

Commit:
Merge branch 'TspMig-healthbot' of https://github.com/welovej/azure-rest-api-specs into TspMig-healthbot

MaryGao commented on specification/healthbot/Healthbot.Management/models.tsp:
```diff
@@ -0,0 +1,460 @@
+   * The error object.
+   */
+  #suppress "@azure-tools/typespec-client-generator-core/property-name-conflict" "FIXME: Update justification, follow aka.ms/tsp/conversion-fix for details"
+  error?: {
```
Can we directly give a name for this anonymous model as m4 name? Here would be ErrorError?

Since we agreed we would report errors in TCGC side if any anonymous models are disallowed?

welovej committed:
update ErrorError

v-jiaodi committed:
Merge branch 'main' into TspMig-healthbot

welovej committed:
prettier

Commit:
update

welovej committed:
remove csharp emmiter

welovej committed:
Rename Healthbot.Management to HealthBot.Management

All files in the specification/healthbot/Healthbot.Management directory were renamed to specification/healthbot/HealthBot.Management to standardize the casing of the directory name.

catalinaperalta reviewed (commented):
Things I noticed:
- We still have several places where we're using private decorators. 
- Some of the missing justifications are clearly for back compat, we should double check where we dont need more follow up from the service team and just update the message ourselves
- There's inconsistent casing for the file names

pshao25 commented on specification/healthbot/HealthBot.Management/models.tsp:
```diff
@@ -0,0 +1,463 @@
+/**
+ * The error object.
+ */
+model ErrorError {
```
This is the context: https://github.com/Azure/azure-rest-api-specs/pull/34069#discussion_r2196693709

Previously it is an anonymous model, which is not recommended, so we follow the name from M4.

catalinaperalta replied:
Could we just name is Error?

pshao25 replied:
It would be a breaking change for SDK if we change the model name.

pshao25 commented on specification/healthbot/HealthBot.Management/main.tsp:
```diff
@@ -0,0 +1,72 @@
+import "@typespec/versioning";
+import "@azure-tools/typespec-azure-core";
+import "@azure-tools/typespec-azure-resource-manager";
+import "./models.tsp";
```
@welovej Please remove the model import in healthbot.tsp.

welovej replied:
removed

pshao25 commented on specification/healthbot/HealthBot.Management/client.tsp:
Here is our guidance: https://azure.github.io/typespec-azure/docs/migrate-swagger/checklists/migrate-arm-tips/

✅ DO add customizations that impact only generated client SDKs in client.tsp

✅ DO add customizations that impact both generated client SDKs and generated OpenAPI specs in back-compat.tsp

pshao25 commented on specification/healthbot/HealthBot.Management/HealthBot.tsp:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * Delete a HealthBot.
+   */
+  #suppress "@azure-tools/typespec-azure-resource-manager/arm-delete-operation-response-codes" "FIXME: Update justification, follow aka.ms/tsp/conversion-fix for details"
```
This is Laurant gave me in wave 1.

pshao25 commented on specification/healthbot/HealthBot.Management/HealthBot.tsp:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * SKU of the Azure Health Bot.
+   */
+  #suppress "@azure-tools/typespec-azure-resource-manager/arm-resource-invalid-envelope-property" "FIXME: Update justification, follow aka.ms/tsp/conversion-fix for details"
```
The envelop property in our library is 
```
@doc("The SKU (Stock Keeping Unit) assigned to this resource.")
model ResourceSkuProperty {
  @doc("The SKU (Stock Keeping Unit) assigned to this resource.")
  sku?: Sku;
}
```
However, in the original swagger, the sku is required so that we cannot use the envelop one.

> Can we design this differently?

Sounds a question to @markcowl

welovej committed:
fix for comment

welovej committed:
remove private things

welovej committed:
fix final-stat-via

welovej committed:
Merge branch 'main' into TspMig-healthbot

pshao25 commented on specification/healthbot/HealthBot.Management/main.tsp:
```diff
@@ -0,0 +1,72 @@
+   * The 2025-05-25 API version.
+   */
+  @useDependency(Azure.ResourceManager.Versions.v1_0_Preview_1)
+  @useDependency(Azure.Core.Versions.v1_0_Preview_1)
```
@welovej remove these two lines.

welovej replied:
removed

welovej commented on specification/healthbot/HealthBot.Management/HealthBot.tsp:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * Patch a HealthBot.
+   */
+  #suppress "@azure-tools/typespec-azure-core/response-schema-problem" "FIXME: Update justification, follow aka.ms/tsp/conversion-fix for details"
```
removed

welovej committed:
update

guy-microsoft commented:
@welovej Is it ready for review by the service team?

welovej commented:
> @welovej Is it ready for review by the service team?

teams contacted.

welovej committed:
update new config for golang

welovej committed:
Update main.tsp

welovej committed:
rename ErrorError to ErrorDetail

welovej committed:
Update client.tsp

welovej committed:
update

guy-microsoft reviewed (approved):
Approved by service team.

pshao25 committed:
Merge branch 'main' into TspMig-healthbot

welovej committed:
Remove unused dependency annotations from Versions enum

Removed unused dependency annotations for API version.

welovej committed:
Update emitter output directories in tspconfig.yaml
